### ISSUE TITLE
Invalid MethodSignature mismatch with psalm-consistent-constructor

### ISSUE THREAD

**Issue by danog** [vimeo/psalm#5527]
If psalm-consistent-constructor is used, and somewhere along the class hierarchy a class declares more constructor arguments than in the root class (while marking them as optional, so psalm-consistent-constructor is not violated), all subsequent subclasses will complain about signature mismatches: https://psalm.dev/r/f87e630c40

**Comment by weirdan** [comment]
Is it invalid though? What if we have a function like this:

```php
/** @param class-string<b> $classname */
function make(string $classname): b {
   return new $classname("x", "y");
}
make(c::class);
```

**Comment by danog** [comment]
Good point, I really should change the design of that specific class hierarchy.

On the other hand, I specified psalm-consistent-constructor on class `a`, not on class `b`; I don't care if someone incorrectly constructs a `class-string<b>`, I only really care that everyone uses `class-string<a>` correctly; idk if that ought to be accounted for inside of psalm (maybe with a variation like psalm-consistent-constructor-no-hierarchy).

**Comment by weirdan** [comment]
What I think `consistent-constructor` should do is to make descendant constructors follow LSP, just like normal methods. And you can't drop optional arguments for those: https://3v4l.org/qbSIa

**Comment by danog** [comment]
Yeah, noticed that quirk when writing phabel)))
On its own, a psalm annotation to force LSP-respecting constructors is a very good idea, just consider giving an option to enable enforcing LSP only when comparing to the current method (I might get around to implementing that, if I actually find a good use-case that just can't be refactored))
